src/connect.ts

```diff
@@ -23,178 +22,144 @@ export type MapDispatchToPropsFn<S, P, OP> = (dispatch: Dispatch<S>, ownProps: O
- * @param {(props: (SP & DP)) => TemplateResult} templateFn The 🔥-html templating function.
- * @returns {FitElement<S, SP & DP, OP>} A newly created 💪-element class.
+ * @param {MapStateToPropsFn<S, SP, OP>} mapStateToProps The MapStateToProps function or factory.
+ * @param {MapDispatchToPropsFn<S, DP, OP>} mapDispatchToProps The MapStateToDispatch function.
+ * @returns The decorator function.
  * @template S, SP, DP, OP
  */
 export default function connect<S, SP, DP, OP = {}>(
     mapStateToProps: MapStateToPropsFactory<S, SP, OP> | MapStateToPropsFn<S, SP, OP>,
     mapDispatchToProps: MapDispatchToPropsFn<S, DP, OP> | DP,
-    templateFn: (props: SP & DP) => TemplateResult,
-): ClassConstructor<FitElement<S, SP & DP, OP>> {
-    return class extends HTMLElement {
-        private _isConnected: boolean = false;
-        private _nodeName: string = this.nodeName.toLowerCase();
-        private _preparedDispatch: MapDispatchToPropsFn<S, DP, OP> | ActionCreatorsMapObject;
-        private _preparedMapStateToProps: MapStateToPropsFn<S, SP, OP>;
-        private _previousProps: SP & DP | null = null;
-        private _renderEnqueued: boolean = false;
-        private _store: Store<S>;
-        private _unsubscribe: Unsubscribe;
-
-        constructor() {
-            super();
-
-            this.attachShadow({ mode: 'open' });
-
-            this._preparedMapStateToProps = isFactory(mapStateToProps)
-                ? mapStateToProps()
-                : mapStateToProps;
-        }
-
-        connectedCallback() {
-            this._isConnected = true;
-
-            const store = this.getStore();
-            this._preparedDispatch = isFunction(mapDispatchToProps)
-                ? mapDispatchToProps
-                : bindActionCreators(mapDispatchToProps as any as ActionCreatorsMapObject, store.dispatch);
-            this._unsubscribe = store.subscribe(() => this.enqueueRender());
-
-            if (window.ShadyCSS) {
-                window.ShadyCSS.styleElement(this);
-            }
-            this.enqueueRender();
-        }
+): TemplateDecorator<S, SP, DP, OP> & ElementDecorator<S, SP, DP, OP>;
 
-        disconnectedCallback() {
-            this._isConnected = false;
+// We declare the exported typings separately for now (above).
 
-            this._unsubscribe();
-            this._store = undefined!;
-        }
+export default function connect<S, SP, DP, OP = {}>(
+    mapStateToProps: MapStateToPropsFactory<S, SP, OP> | MapStateToPropsFn<S, SP, OP>,
+    mapDispatchToProps: MapDispatchToPropsFn<S, DP, OP> | DP,
+) {
+    return <
+        B extends ClassConstructor<HTMLElement>,
+        T extends FitDecorated<B, OP, SP & DP>,
+    >(b: T | TemplateFunction<SP & DP>) => {
+        const base: ClassConstructor<any> = isTemplateFunction(b) ? withFit(b)(HTMLElement) : b;
```

Comment:
Is there any way to check if `b` is already decorated? Otherwise stuff like `getStore` might fail.

Reply:
Nevermind this...

Reply:
Oh, but `super.ownProps` may not exist!

Reply:
If b has already gone through `connect` once, `getStore` will check for existence of `super.getStore` and use that instead, so things should continue to work. :)

Reply:
> Oh, but `super.ownProps` may not exist!

Yes it will, because we are either dealing with something that's already FitDecorated or will be FitDecorated by ourselves, no?